File: haskell/day12_part2_2016.hs
import System.IO

data Instruction = Cpy String String | Inc String | Dec String | Jnz String String deriving (Show)
type Registers = (Int, Int, Int, Int)

main = do
    contents <- readFile "input.txt"
    let instructions = map parseInstruction (lines contents)
    let registers = execute instructions (0, 0, 1, 0) 0
    print $ getRegister registers "a"

parseInstruction :: String -> Instruction
parseInstruction line =
    let parts = words line
    in case parts of
        ["cpy", x, y] -> Cpy x y
        ["inc", x] -> Inc x
        ["dec", x] -> Dec x
        ["jnz", x, y] -> Jnz x y

execute :: [Instruction] -> Registers -> Int -> Registers
execute instructions registers pc
    | pc < 0 || pc >= length instructions = registers
    | otherwise =
        case instructions !! pc of
            Cpy x y -> execute instructions (setRegister registers y (getValue registers x)) (pc + 1)
            Inc x -> execute instructions (setRegister registers x (getRegister registers x + 1)) (pc + 1)
            Dec x -> execute instructions (setRegister registers x (getRegister registers x - 1)) (pc + 1)
            Jnz x y ->
                if getValue registers x /= 0
                then execute instructions registers (pc + read y)
                else execute instructions registers (pc + 1)

getValue :: Registers -> String -> Int
getValue registers x
    | x == "a" = getRegister registers "a"
    | x == "b" = getRegister registers "b"
    | x == "c" = getRegister registers "c"
    | x == "d" = getRegister registers "d"
    | otherwise = read x

getRegister :: Registers -> String -> Int
getRegister (a, b, c, d) x
    | x == "a" = a
    | x == "b" = b
    | x == "c" = c
    | x == "d" = d

setRegister :: Registers -> String -> Int -> Registers
setRegister (a, b, c, d) x value
    | x == "a" = (value, b, c, d)
    | x == "b" = (a, value, c, d)
    | x == "c" = (a, b, value, d)
    | x == "d" = (a, b, c, value)

File: r/day12_part2_2016.R

# Function to execute assembunny code
execute_assembunny <- function(instructions, initial_registers = c(a = 0, b = 0, c = 0, d = 0)) {
  registers <- initial_registers
  instruction_pointer <- 1
  
  while (instruction_pointer >= 1 && instruction_pointer <= length(instructions)) {
    instruction <- instructions[instruction_pointer]
    parts <- strsplit(instruction, " ")[[1]]
    
    opcode <- parts[1]
    
    if (opcode == "cpy") {
      source <- parts[2]
      dest <- parts[3]
      
      if (grepl("^[a-d]$", source)) {
        value <- registers[source]
      } else {
        value <- as.integer(source)
      }
      registers[dest] <- value
      instruction_pointer <- instruction_pointer + 1
      
    } else if (opcode == "inc") {
      reg <- parts[2]
      registers[reg] <- registers[reg] + 1
      instruction_pointer <- instruction_pointer + 1
      
    } else if (opcode == "dec") {
      reg <- parts[2]
      registers[reg] <- registers[reg] - 1
      instruction_pointer <- instruction_pointer + 1
      
    } else if (opcode == "jnz") {
      check <- parts[2]
      jump <- parts[3]
      
      if (grepl("^[a-d]$", check)) {
        check_value <- registers[check]
      } else {
        check_value <- as.integer(check)
      }
      
      if (check_value != 0) {
        instruction_pointer <- instruction_pointer + as.integer(jump)
      } else {
        instruction_pointer <- instruction_pointer + 1
      }
    }
  }
  return(registers)
}

# Read instructions from input.txt
instructions <- readLines("input.txt")

# Part 1: Execute with initial registers all 0
registers_part1 <- execute_assembunny(instructions)
print(paste("Part 1: Value in register a =", registers_part1["a"]))

# Part 2: Execute with register c initialized to 1
registers_part2 <- execute_assembunny(instructions, initial_registers = c(a = 0, b = 0, c = 1, d = 0))
print(paste("Part 2: Value in register a =", registers_part2["a"]))
